crates/sui-core/src/epoch/reconfiguration.rs

```diff
@@ -0,0 +1,135 @@
+use typed_store::Map;
+
+// TODO: Make last checkpoint number of each epoch more flexible.
+pub const CHECKPOINT_COUNT_PER_EPOCH: u64 = 200;
+
+const MAX_START_EPOCH_WAIT_SECONDS: Duration = Duration::from_secs(5);
+
+impl<A> ActiveAuthority<A> {
+    pub async fn start_epoch_change(&self) -> SuiResult {
+        {
+            let checkpoints = self.state.checkpoints.as_ref().unwrap().lock();
+            let next_cp = checkpoints.get_locals().next_checkpoint;
+            fp_ensure!(
+                Self::is_second_last_checkpoint_epoch(next_cp),
+                SuiError::InconsistentEpochState {
+                    error: "start_epoch_change called at the wrong checkpoint".to_owned(),
+                }
+            );
+            fp_ensure!(
+                checkpoints.lowest_unprocessed_checkpoint() == next_cp,
+                SuiError::InconsistentEpochState {
+                    error:
+                        "start_epoch_change called when there are still unprocessed transactions"
+                            .to_owned(),
+                }
+            );
+            // drop checkpoints lock
+        }
+
+        self.state.halted.store(true, Ordering::SeqCst);
+        let instant = Instant::now();
+        while !self.state.batch_notifier.ticket_drained() {
+            tokio::time::sleep(Duration::from_millis(10)).await;
+            fp_ensure!(
+                instant.elapsed() <= MAX_START_EPOCH_WAIT_SECONDS,
+                SuiError::InconsistentEpochState {
+                    error:
+                        "Waiting for batch_notifier ticket to drain timed out in start_epoch_change"
+                            .to_owned(),
+                }
+            );
+        }
+        Ok(())
+    }
+
+    pub async fn finish_epoch_change(&self) -> SuiResult {
+        fp_ensure!(
+            self.state.halted.load(Ordering::SeqCst),
+            SuiError::InconsistentEpochState {
+                error: "finish_epoch_change called when validator is not halted".to_owned(),
+            }
+        );
+        {
+            let checkpoints = self.state.checkpoints.as_ref().unwrap().lock();
+            let next_cp = checkpoints.get_locals().next_checkpoint;
+            fp_ensure!(
+                Self::is_last_checkpoint_epoch(next_cp),
+                SuiError::InconsistentEpochState {
+                    error: "finish_epoch_change called at the wrong checkpoint".to_owned(),
+                }
+            );
+            fp_ensure!(
+                checkpoints.lowest_unprocessed_checkpoint() == next_cp,
+                SuiError::InconsistentEpochState {
+                    error:
+                        "finish_epoch_change called when there are still unprocessed transactions"
+                            .to_owned(),
+                }
+            );
+            if checkpoints.extra_transactions.iter().next().is_some() {
+                // TODO: Revert any tx that's executed but not in the checkpoint.
```

Comment:
pro tip: here the `extras` table in the checkpoint db can help you work out (after the last checkpoint) which transactions to revert.